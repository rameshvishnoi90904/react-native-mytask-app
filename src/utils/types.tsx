export type task = {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export type dispatchActionType = {
  type: string;
  data: task | null;
  searchInput: string;
}

export type NavigationScreenPropsType = {
  Home: undefined;
  TodoScreen: {item: task | null };
  LandingScreen: {};
};
