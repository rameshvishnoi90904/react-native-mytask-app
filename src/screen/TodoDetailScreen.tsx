import React from "react"

const TodoDetailScreen = () => {
    return (
        <></>
    )
}
export default TodoDetailScreen;