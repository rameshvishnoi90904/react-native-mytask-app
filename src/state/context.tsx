
import { createContext } from 'react';
import { dispatchActionType, task } from '../utils/types';

export type AppContextType = {
    tasks: task[];
    dispatch: (action: dispatchActionType) => void;
}

export const AppContext = createContext<AppContextType | null>(null);