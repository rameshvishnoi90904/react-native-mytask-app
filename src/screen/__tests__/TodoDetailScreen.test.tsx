import { render, screen, fireEvent } from '@testing-library/react-native';
import {it, jest, expect} from '@jest/globals';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { AppContext } from '../../state/context';
import { NavigationScreenPropsType, dispatchActionType, task } from '../../utils/types';
import TodoDetailScreen from '../TodoDetailScreen';
import { Alert } from 'react-native';
const Stack = createNativeStackNavigator<NavigationScreenPropsType>();


it("renders correctly", () => {

    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const tree = render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: mockTasks[0]}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    ).toJSON()

    expect(tree).toMatchSnapshot();
})

it('renders correct task', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const openedTask = mockTasks[0]

    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )
    
    expect(screen.getByText(openedTask.title)).toBeTruthy();
    expect(screen.getByText(openedTask.body)).toBeTruthy();
})

it('trigger delete', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const openedTask = mockTasks[0]

    jest.spyOn(Alert, 'alert');


    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )

    fireEvent.press(screen.getByText("Delete"))
    expect(Alert.alert).toHaveBeenCalledTimes(1)
})

it('triggers edit mode ', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const openedTask = mockTasks[0]

    jest.spyOn(Alert, 'alert');


    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )
    fireEvent.press(screen.getByText("Edit"))
    expect(screen.getByText('Save')).toBeTruthy()
})

it('triggers save button ', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const openedTask = mockTasks[0]

    jest.spyOn(Alert, 'alert');


    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )
    fireEvent.press(screen.getByText("Edit"))
    fireEvent.press(screen.getByText("Save"))
    expect(Alert.alert).toBeCalledTimes(1)
})


it('triggers update action ', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    const openedTask = mockTasks[0]

    jest.spyOn(Alert, 'alert');


    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )
    fireEvent.press(screen.getByText("Edit"))
    const updatedTitle = "Update task title triggered by testcase"
    fireEvent.changeText(screen.getByPlaceholderText("Enter your Title here"), updatedTitle)
    
    fireEvent.press(screen.getByText("Save"))

    expect(mockDispatch).toBeCalledWith({type: "update", searchInput: "", data:{...openedTask, title: updatedTitle}})
})




it('triggers revert action ', async () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()
    const openedTask = mockTasks[0]
    jest.spyOn(Alert, 'alert');

    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen} initialParams={{item: openedTask}}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )
    fireEvent.press(screen.getByText("Edit"))
    const updatedTitle = "Update task title triggered by testcase"
    fireEvent.changeText(screen.getByPlaceholderText("Enter your Title here"), updatedTitle)
    
    await fireEvent.press(screen.getByText("Revert"))
    // fireEvent.press(screen.getByText("Yes"))
    // expect(Alert.alert).toBeCalledTimes(1)
    // expect(screen.getByPlaceholderText("Enter your Title here")).toEqual(openedTask.title)
})