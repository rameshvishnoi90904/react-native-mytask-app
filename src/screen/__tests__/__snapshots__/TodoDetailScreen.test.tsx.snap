// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <RCTSafeAreaView
            style={
              {
                "backgroundColor": "#EFEBE0",
                "flex": 1,
              }
            }
          >
            <View
              style={
                {
                  "flex": 1,
                  "paddingHorizontal": 8,
                }
              }
            >
              <View
                style={
                  {
                    "marginVertical": 16,
                  }
                }
              >
                <Text
                  style={
                    {
                      "fontSize": 15,
                    }
                  }
                >
                  Title
                </Text>
                <View
                  style={
                    {
                      "borderColor": "#DCDCDC",
                      "borderRadius": 8,
                      "borderWidth": 1,
                      "marginTop": 8,
                      "paddingHorizontal": 8,
                      "paddingVertical": 12,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#333333",
                        "fontSize": 20,
                        "fontWeight": "bold",
                        "paddingTop": 0,
                      }
                    }
                  >
                    sample title for testing
                  </Text>
                </View>
              </View>
              <RCTScrollView
                contentContainerStyle={
                  {
                    "flex": 1,
                    "marginBottom": 12,
                  }
                }
                style={
                  {
                    "flex": 1,
                  }
                }
              >
                <View>
                  <Text
                    style={
                      {
                        "fontSize": 15,
                      }
                    }
                  >
                    Body
                  </Text>
                  <View
                    style={
                      {
                        "borderColor": "#DCDCDC",
                        "borderRadius": 8,
                        "borderWidth": 1,
                        "marginTop": 8,
                        "paddingHorizontal": 8,
                        "paddingVertical": 12,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#333333",
                          "paddingTop": 0,
                        }
                      }
                    >
                      sample body for testing
                    </Text>
                  </View>
                </View>
              </RCTScrollView>
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-around",
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#FB4570",
                      "borderRadius": 8,
                      "flex": 1,
                      "justifyContent": "center",
                      "marginHorizontal": 6,
                      "paddingVertical": 12,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#EFEBE0",
                        "fontSize": 18,
                      }
                    }
                  >
                    Edit
                  </Text>
                </View>
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#FB4570",
                      "borderRadius": 8,
                      "flex": 1,
                      "justifyContent": "center",
                      "marginHorizontal": 6,
                      "paddingVertical": 12,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#EFEBE0",
                        "fontSize": 18,
                      }
                    }
                  >
                    Delete
                  </Text>
                </View>
              </View>
            </View>
          </RCTSafeAreaView>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="TodoScreen"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
